File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-java/src/main/java/com/sdkwork/Mail/standard/mailStandardContract.java
package com.sdkwork.Mail.standard;

public final class MailStandardContract {

  private MailStandardContract() {
  }

  public interface MailProviderDriver<TNativeClient> {
    String providerKey();

    MailClient<TNativeClient> createClient();
  }

  public interface MailDriverManager<TNativeClient> {
    MailProviderDriver<TNativeClient> resolveDriver(String providerKey);
  }

  public interface MailDataSource<TNativeClient> {
    MailClient<TNativeClient> createClient();
  }

  public interface MailClient<TNativeClient> {
    void connectTransport();

    void authenticateTransport();

    void disconnectTransport();

    void sendMail();

    void probeMailbox();

    void syncMailbox();

    void healthCheck();

    TNativeClient unwrap();
  }

  public interface MailRuntimeController<TNativeClient> {
    void connectTransport();

    void authenticateTransport();

    void disconnectTransport();

    void sendMail();

    void probeMailbox();

    void syncMailbox();

    void healthCheck();
  }
}

File: sdks/sdkwork-mail-sdk/sdkwork-mail-sdk-kotlin/src/main/kotlin/com/sdkwork/Mail/standard/mailStandardContract.kt
package com.sdkwork.Mail.standard

object MailStandardContract {
    const val SYMBOL: String = "MailStandardContract"
}

interface MailProviderDriver<TNativeClient> {
    val providerKey: String

    fun createClient(): MailClient<TNativeClient>
}

interface MailDriverManager<TNativeClient> {
    fun resolveDriver(providerKey: String): MailProviderDriver<TNativeClient>
}

interface MailDataSource<TNativeClient> {
    fun createClient(): MailClient<TNativeClient>
}

interface MailClient<TNativeClient> {
    fun connectTransport()

    fun authenticateTransport()

    fun disconnectTransport()

    fun sendMail()

    fun probeMailbox()

    fun syncMailbox()

    fun healthCheck()

    fun unwrap(): TNativeClient?
}

interface MailRuntimeController<TNativeClient> {
    fun connectTransport()

    fun authenticateTransport()

    fun disconnectTransport()

    fun sendMail()

    fun probeMailbox()

    fun syncMailbox()

    fun healthCheck()
}
